File: lintcode/71.binary-tree-zigzag-level-order-traversal.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
 
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: A list of lists of integer include 
     *          the zigzag level order traversal of its nodes' values 
     */
    public ArrayList<ArrayList<Integer>> zigzagLevelOrder(TreeNode root) {
        // write your code here
        Deque<TreeNode> dq = new LinkedList<>();
        ArrayList<ArrayList<Integer>> res = new ArrayList<>();
        if(root==null) return res;
        res.add(new ArrayList<Integer>(Arrays.asList(root.val)));
        dq.offer(root);
        int countc=1, countn=0;
        boolean reverse = true;
        while(!dq.isEmpty()) {
            if(countc>0) {
                TreeNode cur = dq.pop();
                countc--;
                if(cur.left!=null) {
                    dq.offer(cur.left);
                    countn++;
                }
                if(cur.right!=null) {
                    dq.offer(cur.right);
                    countn++;
                }
            } else {
                // print next level first
                res.add(new ArrayList<Integer>());
                ArrayList<Integer> curList = res.get(res.size()-1);
                Iterator<TreeNode> iter;
                if(reverse) {
                    iter = dq.descendingIterator();
                } else {
                    iter = dq.iterator();
                }
                while(iter.hasNext()){
                    curList.add(iter.next().val);
                }
                countc=countn;
                countn=0;
                reverse=!reverse;
            }
        }
        return res;
    }
}

File: lintcode/71.binary-tree-zigzag-level-order-traversal.cc
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
 
 
class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: A list of lists of integer include
     *          the zigzag level order traversal of its nodes' values
     */
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode *root) {
        // write your code here
        deque<TreeNode*> dq;
        vector<vector<int>> res;
        if(!root) return res;
        int countc=1, countn=0;
        dq.push_back(root);
        res.push_back(vector<int>(1, root->val));
        bool reverse=true;
        while(!dq.empty()) {
            if(countc>0) {
                TreeNode* cur=dq.front();
                dq.pop_front();
                if(cur->left) {
                    dq.push_back(cur->left);
                    countn++;
                }
                if(cur->right) {
                    dq.push_back(cur->right);
                    countn++;    
                }
                countc--;
            } else {
                vector<int> tmp;
                if(reverse) {
                    for(auto itr=dq.rbegin(); itr!=dq.rend(); itr++) {
                        tmp.push_back((*itr)->val);
                    }
                } else {
                    for(auto itr=dq.begin(); itr!=dq.end(); itr++) {
                        tmp.push_back((*itr)->val);
                    }
                }
                res.push_back(tmp);
                countc=countn;
                reverse=!reverse;
                countn=0;
            }
        }
        return res;
    }
};
